### PR TITLE
20240810 application passwords

### LINKED ISSUE #41: Add support for API-application passwords
It's common that applications (like email, imap etc), need passwords and can't supply 2fa. Allow generation of passwords for these services, and allow them to have constrained rights related to their operations and interactions.

### PR DESCRIPTION
**Pull request by Firstyear**
Supersedes #2578 Relates #41 

This rebases the PR to latest master, and I'll do the remaining review items. 

Checklist

- [ x ] This PR contains no AI generated code
- [ x ] `cargo fmt` has been run
- [ ] `cargo clippy` has been run
- [ x ] `cargo test` has been run and passes
- [ ] book chapter included (if relevant)
- [ ] design document included (if relevant)

### CODE REVIEW

File: server/lib/src/constants/groups.rs [hunk 1]
```diff
@@ -407,6 +407,15 @@ lazy_static! {
         ],
         ..Default::default()
     };
+
+    pub static ref BUILTIN_GROUP_APPLICATION_ADMINS: BuiltinGroup = BuiltinGroup {
+        name: "idm_ldap_application_admins",
```
Comment (yaleman) [inline server/lib/src/constants/groups.rs:412]: just so it's consistent TBH, can we make sure it matches across similar things like `UUID_IDM_APPLICATION_ADMINS` which doesn't match its group name
```suggestion
    pub static ref BUILTIN_GROUP_LDAP_APPLICATION_ADMINS: BuiltinGroup = BuiltinGroup {
        name: "idm_ldap_application_admins",
```

File: server/lib/src/server/delete.rs [hunk 2]
```diff
@@ -115,6 +115,15 @@ impl<'a> QueryServerWriteTransaction<'a> {
         {
             self.changed_flags.insert(ChangeFlag::ACP)
         }
+
+        if !self.changed_flags.contains(ChangeFlag::APPLICATION)
```
Comment (Firstyear) [inline server/lib/src/server/delete.rs:119]: doubl check

File: server/lib/src/server/mod.rs [hunk 3]
```diff
@@ -978,6 +981,15 @@ pub trait QueryServerTransaction<'a> {
         )))
     }
 
+    fn get_ldap_applications_set(
```
Comment (Firstyear) [inline server/lib/src/server/mod.rs:984]: rename

File: server/lib/src/server/modify.rs [hunk 4]
```diff
@@ -415,6 +424,16 @@ impl<'a> QueryServerWriteTransaction<'a> {
         {
             self.changed_flags.insert(ChangeFlag::ACP)
         }
+
```
Comment (Firstyear) [inline server/lib/src/server/modify.rs:427]: double check

**Review by Firstyear — COMMENTED** [review]
Self review notes.

**Comment by Firstyear** [comment]
This also fixes all the clippy lints. After this, we should make it a hard requirement.

File: server/lib/src/idm/event.rs [hunk 5]
```diff
@@ -276,6 +276,26 @@ impl LdapTokenAuthEvent {
...
+        usr_uuid: Uuid,
+        cleartext: String,
+    ) -> Result<Self, OperationError> {
+        Ok(LdapApplicationAuthEvent {
+            application: app_name.to_string(),
+            target: usr_uuid,
+            cleartext,
+        })
+    }
+}
```
Comment (yaleman) [inline server/lib/src/idm/event.rs:297]: Shouldn't this just be a `new` ?

```suggestion
impl LdapApplicationAuthEvent {
    pub fn new(
        app_name: impl ToString,
        usr_uuid: Uuid,
        cleartext: String,
    ) -> Result<Self, OperationError> {
        Ok(LdapApplicationAuthEvent {
            application: app_name.to_string(),
            target: usr_uuid,
            cleartext,
        })
    }
}
```

**Review by yaleman — CHANGES_REQUESTED** [review]

File: proto/src/internal/error.rs [hunk 6]
```diff
@@ -374,8 +374,8 @@ impl OperationError {
             Self::KP0043KeyObjectJweA128GCMEncryption => None,
             Self::KP0044KeyObjectJwsPublicJwk => None,
             Self::PL0001GidOverlapsSystemRange => None,
-            Self::UI0001ChallengeSerialisation => None,
-            Self::UI0002InvalidState => None,
+            Self::UI0001ChallengeSerialisation => Some("The webauthn challenge was unable to be serialised."),
```
Comment (yaleman) [inline proto/src/internal/error.rs:377]:
```suggestion
            Self::UI0001ChallengeSerialisation => Some("The WebAuthn challenge was unable to be serialised."),
```

**Review by yaleman — APPROVED** [review]

**Pull request merged**